### ISSUE TITLE
Metadata for Typedefs to convert them to Haxe Classes at compile-time

### ISSUE THREAD

**Issue by larsiusprime** [HaxeFoundation/haxe#4526]
Typedefs are an extremely convenient feature, but they get resolved at compile time to generic objects that rely on reflection. This is good for some use cases where that's exactly what you want (particularly if you want to rely on duck typing and/or adding arbitrary properties later), but it would be handy to be able to single out specific Typedefs to be treated more stringently so you can realize the performance gains of static typing. A lot of developers tend to use these in "hot" areas of code and it can really add up.

So something like this perhaps:

```haxe
@:struct
typedef MyStruct = {
  var a:Int;
  var b:String;
}
```

Would become this:

```haxe
class MyStruct
{
  public var a:Int;
  public var b:String;
  public inline new():Void{}
}
```

Basically would let you use typedefs in a way more similar to C++ structs.

**Comment by delahee** [comment]
+1 with all my heart as I spent a tremendous amount of time getting rid of these little crepers especially in other people's code.

**Comment by Chlorodatafile** [comment]
+1, because it can be very usefull for making user friendly thing without drink all the computer capacity !

**Comment by andyli** [comment]
With [dataclass](https://github.com/ciscoheat/dataclass), you can simply do a regexp replace that turn
```
typedef MyStruct = {
  var a:Int;
  var b:String;
}
```
into
```
class MyStruct implements dataclass.DataClass {
  var a:Int;
  var b:String;
}
```

problem solved? ;)

**Comment by delahee** [comment]
@andyli except {a:"foo"} to var t=new(); t.a= "foo" is fairly advanced regexp that I would be happy to ever be written by a human being :)

**Comment by andyli** [comment]
Hey, then that's much more than "converting struct types to classes"! Anyway, one can write build macro that turns build an abstract that has an `@:from macro static function` that accepts a struct expression.

I'm not sure how `implements` is a problem for proper inlining?

**Comment by delahee** [comment]
@andyli I guess it guarantees overriding which we don't want in this situation...after tests it pragmatically cancels some inlining cases.

**Comment by andyli** [comment]
@delahee Would you give a simple example of that? I don't think it should be the case here since for the dataclass lib, the `implements dataclass.DataClass` thing is merely for `@:autoBuild`. The resulting class built by the macro wouldn't have any method that is inherited/overridden.

**Comment by delahee** [comment]
ok I got it, I thought you wanted to generate a real interface.

**Comment by ncannasse** [comment]
What about the other way:

``` 
@:struct class MyStruct {
  public var a:Int;
  public var b:String;
  public inline new(a,?b) {
      this.a = a;
      this.b = b;
  }
}
// would allow to be initialized with:
var m1 : MyStruct = { a : 0 };
var m2 : MyStruct = { a : 0, b : "hello" };
``` 

This will be much more easy for us to support it in the compiler this way, given that anonymous structures are structurally subtyped so it would require many checks at different places in the compiler.

This syntax would still allow to "upgrade" a typedef to a class by only changing its definition, and maybe adding a few types here and there to make sure that constant structures are properly turned into MyStruct, because for instance the following would error:

``` 
var m1 = { a : 0, b : "test };
var m2 : MyStruct = m1; // error !
```

**Comment by ncannasse** [comment]
But yes maybe this can be done already with macros+abstract. We don't want to allocate any structure here. Maybe we could provide such buildmacro-interface in std lib for convenience. Anyone that want to give it a try ? (let's summon @back2dos)

**Comment by larsiusprime** [comment]
@ncannasse This sounds really cool, Nicolas! It would be very convenient if I could use the object notation to assign values to something that resolves as an actual typed class at compile time. That would reduce the amount of time for refactoring typedefs to struct-classes; It doesn't take much time to replace "typedef" with "class", but it takes a lot of work to replace the object notation assignments, especially when duck typing, etc, is being used.

**Comment by delahee** [comment]
That would be reallyyyyy awesome !

**Comment by Chlorodatafile** [comment]
It's really better than the first idea, with all the things that could allow ! That would be really a good thing !

**Comment by francescoagati** [comment]
@larslusprime ylu can try to use lib objectinit For init Objects with this macro you can pass a struct that is use to initialize object properties

**Comment by Simn** [comment]
I must be missing the point here. How is `var m2 : MyStruct = { a : 0, b : "hello" };` more convenient than `var m2 = new MyStruct(0, "hello");`? You still have the verbose class definition and need a constructor that deals with all the fields. Working with optional fields you also run into the problem of nullability for basic types on static targets.

If anything we should support field initialization similar to C#, e.g. `var m2 = new MyStruct { a: 0, b: "hello" };`. This would go really well with #3442 too:

```haxe
@:struct
class MyStruct {
	public var a(default, new):Int;
	public var b(default, new):String;
}

class Main {
	static function main() {
		var m1 = new MyStruct { a: 1 };
		var m2 = new MyStruct { a: 1, b: "foo" };
		//m2.b = "bar"; // error
	}
}
```

This is equivalent to the current

```haxe
@:struct
class MyStruct {
	public var a:Int;
	public var b:String;
	public inline function new() { }
}

class Main {
	static function main() {
		var m1 = new MyStruct();
		m1.a = 1;

		var m2 = new MyStruct();
		m2.a = 1;
		m2.b = "foo";
	}
}
```

which is already generated as

```js
Main.main = function() {
	var m1_a = 1;
	m1_a = 1;
	var m2_b = "foo";
	var m2_a = 1;
	m2_a = 1;
	m2_b = "foo";
};
```

**Comment by Chlorodatafile** [comment]
Personally, if it's like that : 
```haxe
@:struct
class MyStruct {
    public var a(default, new):Int;
    public var b(default, new):String;
}

class Main {
    static function main() {
        var m1 = new MyStruct { a: 1 };
        var m2 = new MyStruct { a: 1, b: "foo" };
        //m2.b = "bar"; // error
    }
}
```

I found that more user friendly, and more usefull when it had to be used as a function argument like this : 
```haxe
class Test {
    static function make(struct:MyStruct) { ... }
    static function main() {
         make(new MyStruct { a: 1, b: "foo" });
    }
}
```

**Comment by ncannasse** [comment]
@Simn the idea is that if you have an already existing code that uses typedefs, simply changing the typedef into a struct class will give you speed benefits on strictly typed platforms without having to change all the places your code uses `{...}`

**Comment by Simn** [comment]
In that case I don't see the advantage of your suggestion over the original one. Why not simply type a `@:struct typedef` as a class and generate the constructor accordingly? That seems better than having to do that part by hand.

**Comment by ncannasse** [comment]
I think my proposal is better since its allow a class to be initialized with a structure, while still allowing to add methods.

**Comment by Simn** [comment]
You can have methods on structure syntax too. Sure it fails during typing, but that's irrelevant if we treat the whole thing as a class anyway.

By the way, it occurred to me that either solution has to deal with execution order. Going with your example:

```haxe
@:struct class MyStruct {
  public var a:Int;
  public var b:String;
  public inline new(a,?b) {
      this.a = a;
      this.b = b;
  }
}

var m1 : MyStruct = { a : call1(), b : call2() }; // ok
var m2 : MyStruct = { b : call1(), a : call2() }; // need temp vars
```

**Comment by Simn** [comment]
Regarding nullability, there's a difference between structures and classes on static targets:

```haxe
class A {
	@:optional public var x:Int;
	public function new() { }
}

typedef B = {
	@:optional public var x:Int;
}

class Main {
	static function main() {
		var a = new A();
		trace(a.x); // 0

		var b:B = {};
		trace(b.x); // null
	}
}
```

This is probably not a major issue but still has to be considered when "upgrading".

**Comment by Simn** [comment]
I have made two commits to https://github.com/HaxeFoundation/haxe/tree/typedef_struct.

[The first one](https://github.com/HaxeFoundation/haxe/commit/5ffd0ae0180fade731271e568aa0da7d0d994e02) is essentially your suggestion and allows manually defining a `@:struct class`. This currently has some issues when users define funny constructors, e.g. when they have extra arguments that are not part of the structure.

[The second one](https://github.com/HaxeFoundation/haxe/commit/d82c3e67ac08dd027c5a5668ae6286195365481d) allows what @larsiusprime originally suggested by turning `@:struct typedef` into a `@:struct class` and generating the constructor. This automatically fixes the funny constructor problem and in fact we should just disallow custom constructors on `@:struct typedef`.

Note that these commits build on each other: My suggestion is an extension of yours which I consider easier to use (just add `@:struct` to your `typedef` and be done with it) and less error-prone (because users don't have to construct a constructor and potentially mess it up).

**Comment by Simn** [comment]
Looks like `@:struct` already has a meaning on C#, so we have to pick a different metadata for this.

**Comment by underscorediscovery** [comment]
I love the idea of stronger typedef constructs especially on cpp target. 
I would also avoid @:struct, c++ has `::cpp::Struct` features, which are literal in c++ semantics. 
`@:physical`? `@:structured` ?

**Comment by ibilon** [comment]
Any reasons to require a metadata and not make this by default?

**Comment by Simn** [comment]
Structures have different semantics than classes. For example, they allow structural subtyping (think `Iterator<T>`).

**Comment by ncannasse** [comment]
Ok, I'm coming back to this since I need it for some pretty complex macro code generation I'm working on atm. Reading the whole thing again I was thinking that maybe the constructor should not be required, but it still opens up a lot more possibilities.

I'll then merge Simon first commit that implements it, we can improve support for it by auto-generating the constructor if it does not exists.

**Comment by ncannasse** [comment]
Ok I did some initial support for it, and used `@:structInit` not to clash with C#.
I'm leaving it open and assigning to @Simn so he can automatic constructor generation when none is found for such classes.

**Comment by delahee** [comment]
Kudos!

2016-02-03 16:08 GMT+01:00 Nicolas Cannasse <notifications@github.com>:

> Ok I did some initial support for it, and used @:structInit not to clash
> with C#.
> I'm leaving it open and assigning to @Simn <https://github.com/Simn> so
> he can automatic constructor generation when none is found for such classes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFoundation/haxe/issues/4526#issuecomment-179284214>
> .
>



-- 
David Elahee

**Comment by larsiusprime** [comment]
Hurray!

**Comment by Simn** [comment]
Fair warning: I have no idea what happens if a `@:structInit` class extends another class.

**Comment by delahee** [comment]
First, this is a very important feature to me, so first and foremost,
Thanks!

As to inheritance, maybe we could then forbid it for now so we avoid edges
cases, seems fair to me...

Xo
Le 21 févr. 2016 12:13, "Simon Krajewski" <notifications@github.com> a
écrit :
